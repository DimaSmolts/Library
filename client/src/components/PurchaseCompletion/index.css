.purchase-completion {
  display: grid;
  grid-gap: 1em 0;
  grid-template-columns: 52% 16% 16% 16%;
  width: 75%;
  margin: 25px auto;

  background: var(--gold);
  padding: 10px;
  box-sizing: border-box;
  border-radius: 15px;
  border: 2px solid var(--dark-blue);
}

.purchase-completion span {
  text-align: center;
  align-self: center;
}

.total {
  grid-column: 1/4;
}

.total-line {
  grid-column: 1/5;
  width: 100%;
  height: 2px;
  background: var(--dark-blue);
}

.final-price{
  grid-column: 4;
}

.submit-purchase {
  font-size: 1em;
  font-weight: bold;
  background: var(--gold);
  border-radius: 40px;
  border: 4px solid var(--dark-blue);
  vertical-align: bottom;
  grid-column: 1/5;
  width: 110px;
  margin: 0 auto;
}

.submit-purchase:hover {
  background: var(--dark-gold);
  cursor: pointer;
}

.purchase-column-title {
  font-weight: bold;
}
