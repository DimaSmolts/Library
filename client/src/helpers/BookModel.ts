export class Book {
  id: number;
  name: string;
  author: string;
  genres: Array<string>;
  description: string;
  rate: number;
  price: number;
  image: string;
}
