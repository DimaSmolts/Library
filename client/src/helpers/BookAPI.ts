import axios from 'axios';
import { Book } from './BookModel'

const baseURL = process.env.REACT_APP_API_URL || 'http://localhost:3000';

const instance = axios.create({
  baseURL: baseURL + '/api',
})

export const bookAPI = {
  async getBooks() {
    return await instance.get<Array<Book>>('/books');
  },
  async getBook(id: number) {
    return await instance.get<Book>('/books/' + id);
  }
}
