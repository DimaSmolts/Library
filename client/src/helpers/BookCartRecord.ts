import { Book } from './BookModel'

export interface BookCartRecord {
  book: Book,
  quantity: number,
}
