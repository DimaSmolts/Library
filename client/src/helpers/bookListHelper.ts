import { Book } from './BookModel'

export function filterBookList(books: Array<Book>, query: string): Array<Book> {
  const temp = books.filter(book => {
    return book.name
      .trim()
      .toLowerCase()
      .includes(query.trim().toLowerCase())
  })
  return temp;
}
