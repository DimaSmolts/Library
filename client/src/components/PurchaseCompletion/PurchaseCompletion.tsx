import React from 'react'
import './index.css'
import { Book } from '../../helpers/BookModel'
import { BookCartRecord } from '../../store/reducers/cartReducer'

interface PurchaseCompletionProps {
  bookCartRecords: Array<BookCartRecord>,
  handlePurchaseCompletion: (event: React.FormEvent<EventTarget>) => void,
  addBookToCart: (book: Book) => void,
  removeBookFromCart: (book: Book) => void,
}

const PurchaseCompletion: React.SFC<PurchaseCompletionProps> = (props: PurchaseCompletionProps) => {
  return <div className='purchase-completion'>
    <span className='purchase-column-title'>Book</span>
    <span className='purchase-column-title'>Price</span>
    <span className='purchase-column-title'>Count</span>
    <span className='purchase-column-title'>Sum</span>
    {
      props.bookCartRecords
        .map((br, index) => {
          return <>
            <h2>{index + 1}. {br.book.name}</h2>
            <span>{br.book.price}£</span>
            <span>x{br.quantity}</span>
            <span>{br.quantity * br.book.price}£</span>
          </>
        })
    }
    <div className='total-line'></div>
    <h2 className='total'> Total </h2>
    <span className='final-price'>
      {
        props.bookCartRecords
          .map(br => br.quantity * br.book.price)
          .reduce((total, current) => {
            return total + current;
          })
      }£
      </span>
    <button
      className='submit-purchase'
      onClick={props.handlePurchaseCompletion}>
      Submit</button>
  </div>
}


export default PurchaseCompletion
