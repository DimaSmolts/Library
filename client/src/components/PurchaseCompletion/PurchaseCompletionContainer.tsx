import React from 'react'
import { AppState } from '../../store/rootReducer'
import { connect, ConnectedProps } from 'react-redux';
import { addBookToCartThunk, removeBookFromCartThunk } from '../../store/thunks/cartThunks';
import { Book } from '../../helpers/BookModel';
import PurchaseCompletion from './PurchaseCompletion';
import Message, { MessageTypes, MessageTextSizes } from '../Message/Message';

class PurchaseCompletionContainer extends React.Component<PropsFromRedux> {

  handlePurchaseCompletion(event: React.FormEvent<EventTarget>) {
    alert('Thanks!');
  }

  render() {
    return this.props.bookCartRecords.length === 0 ?
      <Message type={MessageTypes.msgInfo} size={MessageTextSizes.msgBig} text='Nothing added to the cart' />
      :
      <PurchaseCompletion
        bookCartRecords={this.props.bookCartRecords}
        handlePurchaseCompletion={this.handlePurchaseCompletion}
        addBookToCart={this.props.addBookToCart}
        removeBookFromCart={this.props.removeBookFromCart} />
  }
}

const mapState = (state: AppState) => ({
  bookCartRecords: state.cartReducer.bookRecords
})

const mapDispatch = {
  addBookToCart: (book: Book) => addBookToCartThunk(book),
  removeBookFromCart: (book: Book) => removeBookFromCartThunk(book),
}

const connector = connect(mapState, mapDispatch);
type PropsFromRedux = ConnectedProps<typeof connector>

export default connector(PurchaseCompletionContainer)
