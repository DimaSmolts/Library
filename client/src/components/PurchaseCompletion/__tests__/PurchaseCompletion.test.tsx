import React from 'react'
import { shallow } from 'enzyme'

import PurchaseCompletion from '../PurchaseCompletion'
import { BookCartRecord } from '../../../helpers/BookCartRecord'

const bookCartRecords = [{ quantity: 2, book: { name: 'any', price: 20 } },] as Array<BookCartRecord>
const handleCompletion = jest.fn()
const addBookToCart = jest.fn()
const removeBookFromCart = jest.fn()

test('snapshot: <PurchaseCompletion /> renders correctly', () => {
  const tree = shallow(
    <PurchaseCompletion
      bookCartRecords={bookCartRecords}
      handlePurchaseCompletion={handleCompletion}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  expect(tree).toMatchSnapshot()
})

it('should submit purchase', () => {
  const wrapper = shallow(
    <PurchaseCompletion
      bookCartRecords={bookCartRecords}
      handlePurchaseCompletion={handleCompletion}
      addBookToCart={addBookToCart}
      removeBookFromCart={removeBookFromCart} />)

  wrapper.find('.submit-purchase').simulate('click')
  expect(handleCompletion).toBeCalled()
})
