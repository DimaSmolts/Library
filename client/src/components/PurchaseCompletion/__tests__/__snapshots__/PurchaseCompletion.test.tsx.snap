// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: <PurchaseCompletion /> renders correctly 1`] = `
<div
  className="purchase-completion"
>
  <span
    className="purchase-column-title"
  >
    Book
  </span>
  <span
    className="purchase-column-title"
  >
    Price
  </span>
  <span
    className="purchase-column-title"
  >
    Count
  </span>
  <span
    className="purchase-column-title"
  >
    Sum
  </span>
  <h2>
    1
    .
    any
  </h2>
  <span>
    20
    £
  </span>
  <span>
    x
    2
  </span>
  <span>
    40
    £
  </span>
  <div
    className="total-line"
  />
  <h2
    className="total"
  >
     Total
  </h2>
  <span
    className="final-price"
  >
    40
    £
  </span>
  <button
    className="submit-purchase"
    onClick={[MockFunction]}
  >
    Submit
  </button>
</div>
`;
